storage/statedb/database_test.go

```diff
@@ -0,0 +1,124 @@
+
+func TestDatabase_Reference(t *testing.T) {
+	memDB := database.NewMemoryDBManager()
+	db := NewDatabaseWithCache(memDB, 128)
+
+	assert.Equal(t, memDB, db.DiskDB())
+	assert.Equal(t, 1, len(db.nodes)) // {} : {}
+
+	db.Reference(childHash, parentHash)
+	assert.Equal(t, 1, len(db.nodes)) // {} : {}
+
+	child := &cachedNode{}
+	parent := &cachedNode{}
+	db.nodes[childHash] = child
+	db.nodes[parentHash] = parent
+
+	// Call Reference after updating db.nodes
+	db.Reference(childHash, parentHash)
+	assert.Equal(t, 3, len(db.nodes))
+	assert.Equal(t, uint64(1), child.parents)
+	assert.Equal(t, uint64(1), parent.children[childHash])
+
+	// Just calling Reference does not have effect
+	db.Reference(childHash, parentHash)
+	assert.Equal(t, 3, len(db.nodes))
+	assert.Equal(t, uint64(1), child.parents)
+	assert.Equal(t, uint64(1), parent.children[childHash])
+}
+
+func TestDatabase_DeReference(t *testing.T) {
+	memDB := database.NewMemoryDBManager()
+	db := NewDatabaseWithCache(memDB, 128)
+	assert.Equal(t, 1, len(db.nodes)) // {} : {}
+
+	db.Dereference(parentHash)
+	assert.Equal(t, 1, len(db.nodes)) // {} : {}
+	assert.Equal(t, uint64(0), db.gcnodes)
+	assert.Equal(t, common.StorageSize(0), db.gcsize)
+
+	child := &cachedNode{}
+	parent := &cachedNode{}
+	db.nodes[childHash] = child
+	db.nodes[parentHash] = parent
+
+	db.Reference(childHash, parentHash)
+	assert.Equal(t, 3, len(db.nodes))
+	assert.Equal(t, uint64(1), child.parents)
+	assert.Equal(t, uint64(1), parent.children[childHash])
+	assert.Equal(t, uint64(0), db.gcnodes)
+	assert.Equal(t, common.StorageSize(0), db.gcsize)
+
+	db.Dereference(parentHash)
+	assert.Equal(t, 1, len(db.nodes))
+	assert.Equal(t, uint64(0), child.parents)
+	assert.Equal(t, uint64(0), parent.children[childHash])
+	assert.Equal(t, uint64(2), db.gcnodes)
+	assert.Equal(t, common.StorageSize(64), db.gcsize)
+}
+
+func TestDatabase_Size(t *testing.T) {
+	memDB := database.NewMemoryDBManager()
+	db := NewDatabaseWithCache(memDB, 128)
+
+	totalMemorySize, preimagesSize := db.Size()
+	assert.Equal(t, common.StorageSize(0), totalMemorySize)
+	assert.Equal(t, common.StorageSize(0), preimagesSize)
+
+	child := &cachedNode{}
+	parent := &cachedNode{}
+	db.nodes[childHash] = child
+	db.nodes[parentHash] = parent
+
+	db.Reference(childHash, parentHash)
+
+	totalMemorySize, preimagesSize = db.Size()
+	assert.Equal(t, common.StorageSize(128), totalMemorySize)
+	assert.Equal(t, common.StorageSize(0), preimagesSize)
+
+	db.preimagesSize += 100
+	totalMemorySize, preimagesSize = db.Size()
+	assert.Equal(t, common.StorageSize(128), totalMemorySize)
+	assert.Equal(t, common.StorageSize(100), preimagesSize)
+}
+
+func TestDatabase_SecureKey(t *testing.T) {
+	memDB := database.NewMemoryDBManager()
+	db := NewDatabaseWithCache(memDB, 128)
+
+	secKey1 := db.secureKey(childHash[:])
+	copiedSecKey := make([]byte, 0, len(secKey1))
+	copy(copiedSecKey, secKey1)
+
+	secKey2 := db.secureKey(parentHash[:])
```

Comment:
@ehnuje Just curious, why this key is secureKey? :)

Reply:
@KimKyungup `secureKey` is used to make a database key for preimage, however, I have no idea why geth committers choose `secureKey` as its name.